arborio/asc_lexer.cpp

```diff
@@ -0,0 +1,407 @@
+#include <string>
+
+#include <arborio/neurolucida.hpp>
+
+#include "asc_lexer.hpp"
+
+namespace arborio {
+
+namespace asc {
+
+std::ostream& operator<<(std::ostream& o, const tok& t) {
+    switch (t) {
+        case tok::lparen:
+            return o << "lparen";
+        case tok::rparen:
+            return o << "rparen";
+        case tok::lt:
+            return o << "lt";
+        case tok::gt:
+            return o << "gt";
+        case tok::comma:
+            return o << "comma";
+        case tok::real:
+            return o << "real";
+        case tok::integer:
+            return o << "integer";
+        case tok::symbol:
+            return o << "symbol";
+        case tok::string:
+            return o << "string";
+        case tok::pipe:
+            return o << "pipe";
+        case tok::eof:
+            return o << "eof";
+        case tok::error:
+            return o << "error";
+    }
+    return o << "unknown";
+}
+
+std::ostream& operator<<(std::ostream& o, const src_location& l) {
+    return o << "(src-location " << l.line << " " << l.column << ")";
+}
+
+std::ostream& operator<<(std::ostream& o, const token& t) {
+    const char* spelling = t.spelling.c_str();
+    if      (t.kind==tok::eof)   spelling = "\\0";
+    else if (t.kind==tok::error) spelling = "";
+    return o << "(token " << t.kind << " \"" << spelling << "\" " << t.loc << ")";
+}
+
+inline bool is_alphanumeric(char c) {
```

Comment:
I wasn't aware of `isalnum`. I have used it, but I keep this wrapper because I now realise that the `char` should be cast to `unsigned char`:
```
return std::isalnum(static_cast<unsigned char>(ch));
```